src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java

```diff
@@ -0,0 +1,134 @@
+package jp.co.soramitsu.sora.crypto;
+
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.List;
+import lombok.NonNull;
+import org.spongycastle.util.Arrays;
+
+public class MerkleTree {
+
+  private MessageDigest digest;
+
+  // [root] [intermediate nodes] [leafs]
+  private byte[][] tree;
+
+  /**
+   * Creates new tree from given leafs with given digest function.
+   * @param digest implementation of the digest algorithm
+   * @param leafs list of leaf nodes
+   */
+  public MerkleTree(@NonNull MessageDigest digest, @NonNull List<byte[]> leafs) {
```

Comment:
Having two constructors and two flows of working with this object seems confusing
Caller can either call constructor and pass `leafs` immediately, or just create object with `digest` only and then use `#create`. Leave only one way of using, and get rid of other.

Reply:
I recommend leaving second one, when object is created with `digest` only.
Thus we separate initialization from behaviour